hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/ReadBufferManager.java

```diff
@@ -299,11 +327,32 @@ private void clearFromReadAheadQueue(final AbfsInputStream stream, final long re
   }
 
   private int getBlockFromCompletedQueue(final AbfsInputStream stream, final long position, final int length,
-                                         final byte[] buffer) {
-    ReadBuffer buf = getFromList(completedReadList, stream, position);
-    if (buf == null || position >= buf.getOffset() + buf.getLength()) {
+                                         final byte[] buffer) throws IOException {
+    ReadBuffer buf = getBufferFromCompletedQueue(stream, position);
+
+    if (buf == null) {
       return 0;
     }
+
+    if (buf.getStatus() == ReadBufferStatus.READ_FAILED) {
+      // Eviction of a read buffer is triggered only when a queue request comes in
+      // and each eviction attempt tries to find one eligible buffer.
+      // Hence there are chances that an old read-ahead buffer with exception is still
+      // available. To prevent new read requests to fail due to such old buffers,
+      // return exception only from buffers that failed within last THRESHOLD_AGE_MILLISECONDS
+      if ((currentTimeMillis() - (buf.getTimeStamp()) < THRESHOLD_AGE_MILLISECONDS)) {
```

Comment:
These parenthesis are confusing.

Reply:
If the buffer was updated with an error in last "Threshold_age_milliseconds", then we return the error to client. 

Can you please help me with the cause of confusion.

Reply:
If this is to avoid the failure for new requests, then instead of checking the time window, why not  reset the buffer status before throwing the exception? Then following new requests will not be affected by the old failure.

Reply:
Aim here is to enforce the read-ahead failure for the threshold time duration (which currently is 30 sec), i.e. any read request for that offset that can be served by the ReadBuffer needs to fail.

Reply:
why 30s? Any way to tune this for test runs?